src/providers/alchemy_provider/provider.rs

```diff
@@ -0,0 +1,252 @@
+use super::contracts::erc20::EthereumErc20;
+use crate::{
+    models::token::{TokenBalance, TokenBalances, TokenMetadata},
+    providers::eth_provider::{
+        error::EthApiError,
+        provider::{EthProviderResult, EthereumProvider},
+    },
+};
+use async_trait::async_trait;
+use auto_impl::auto_impl;
+use eyre::Result;
+use futures::future::join_all;
+use mongodb::bson::doc;
+use reth_primitives::{Address, BlockId, BlockNumberOrTag, Bytes, B256, U256, U64};
+use reth_rpc_types::{
+    serde_helpers::JsonStorageKey, state::StateOverride, txpool::TxpoolContent, BlockOverrides, FeeHistory, Filter,
+    FilterChanges, Header, Index, RichBlock, SyncStatus, Transaction, TransactionReceipt, TransactionRequest,
+};
+
+#[async_trait]
+#[auto_impl(Arc, &)]
+pub trait AlchemyProvider {
+    /// Retrieves the token balances for a given address.
+    async fn token_balances(
+        &self,
+        address: Address,
+        contract_addresses: Vec<Address>,
+    ) -> EthProviderResult<TokenBalances>;
+    /// Retrieves the metadata for a given token.
+    async fn token_metadata(&self, contract_address: Address) -> EthProviderResult<TokenMetadata>;
+    /// Retrieves the allowance for a given token.
+    async fn token_allowance(
+        &self,
+        contract_address: Address,
+        owner: Address,
+        spender: Address,
+    ) -> EthProviderResult<U256>;
+}
+
+#[derive(Debug, Clone)]
+pub struct AlchemyStruct<P: EthereumProvider> {
```

Comment:
I think we need to rename, maybe into `AlchemyDataProvider` to follow the same naming as the eth provider